// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: examples/protos/custom_extension_example/custom_config.proto

package conf

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on CustomConfig with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *CustomConfig) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on CustomConfig with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in CustomConfigMultiError, or
// nil if none found.
func (m *CustomConfig) ValidateAll() error {
	return m.validate(true)
}

func (m *CustomConfig) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for CustomField

	// no validation rules for CustomNumber

	if all {
		switch v := interface{}(m.GetNested()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, CustomConfigValidationError{
					field:  "Nested",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, CustomConfigValidationError{
					field:  "Nested",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetNested()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return CustomConfigValidationError{
				field:  "Nested",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return CustomConfigMultiError(errors)
	}

	return nil
}

// CustomConfigMultiError is an error wrapping multiple validation errors
// returned by CustomConfig.ValidateAll() if the designated constraints aren't met.
type CustomConfigMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m CustomConfigMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m CustomConfigMultiError) AllErrors() []error { return m }

// CustomConfigValidationError is the validation error returned by
// CustomConfig.Validate if the designated constraints aren't met.
type CustomConfigValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e CustomConfigValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e CustomConfigValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e CustomConfigValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e CustomConfigValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e CustomConfigValidationError) ErrorName() string { return "CustomConfigValidationError" }

// Error satisfies the builtin error interface
func (e CustomConfigValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sCustomConfig.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = CustomConfigValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = CustomConfigValidationError{}

// Validate checks the field values on ApplicationConfig with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *ApplicationConfig) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ApplicationConfig with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ApplicationConfigMultiError, or nil if none found.
func (m *ApplicationConfig) ValidateAll() error {
	return m.validate(true)
}

func (m *ApplicationConfig) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetCustomize()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, ApplicationConfigValidationError{
					field:  "Customize",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, ApplicationConfigValidationError{
					field:  "Customize",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetCustomize()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return ApplicationConfigValidationError{
				field:  "Customize",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return ApplicationConfigMultiError(errors)
	}

	return nil
}

// ApplicationConfigMultiError is an error wrapping multiple validation errors
// returned by ApplicationConfig.ValidateAll() if the designated constraints
// aren't met.
type ApplicationConfigMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ApplicationConfigMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ApplicationConfigMultiError) AllErrors() []error { return m }

// ApplicationConfigValidationError is the validation error returned by
// ApplicationConfig.Validate if the designated constraints aren't met.
type ApplicationConfigValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ApplicationConfigValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ApplicationConfigValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ApplicationConfigValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ApplicationConfigValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ApplicationConfigValidationError) ErrorName() string {
	return "ApplicationConfigValidationError"
}

// Error satisfies the builtin error interface
func (e ApplicationConfigValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sApplicationConfig.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ApplicationConfigValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ApplicationConfigValidationError{}

// Validate checks the field values on CustomConfig_NestedConfig with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *CustomConfig_NestedConfig) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on CustomConfig_NestedConfig with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// CustomConfig_NestedConfigMultiError, or nil if none found.
func (m *CustomConfig_NestedConfig) ValidateAll() error {
	return m.validate(true)
}

func (m *CustomConfig_NestedConfig) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Enabled

	// no validation rules for Name

	if len(errors) > 0 {
		return CustomConfig_NestedConfigMultiError(errors)
	}

	return nil
}

// CustomConfig_NestedConfigMultiError is an error wrapping multiple validation
// errors returned by CustomConfig_NestedConfig.ValidateAll() if the
// designated constraints aren't met.
type CustomConfig_NestedConfigMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m CustomConfig_NestedConfigMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m CustomConfig_NestedConfigMultiError) AllErrors() []error { return m }

// CustomConfig_NestedConfigValidationError is the validation error returned by
// CustomConfig_NestedConfig.Validate if the designated constraints aren't met.
type CustomConfig_NestedConfigValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e CustomConfig_NestedConfigValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e CustomConfig_NestedConfigValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e CustomConfig_NestedConfigValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e CustomConfig_NestedConfigValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e CustomConfig_NestedConfigValidationError) ErrorName() string {
	return "CustomConfig_NestedConfigValidationError"
}

// Error satisfies the builtin error interface
func (e CustomConfig_NestedConfigValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sCustomConfig_NestedConfig.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = CustomConfig_NestedConfigValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = CustomConfig_NestedConfigValidationError{}
